src/mono/mono/metadata/icall.c

```diff
@@ -5966,9 +5966,9 @@ check_for_invalid_array_type (MonoType *type, MonoError *error)
 	gboolean allowed = TRUE;
 	char *name;
 
-	if (m_type_is_byref (type))
+	if (MONO_TYPE_IS_VOID (type))
```

Comment:
I think this only handles reflection. For IL loading we either need a check in `mono_metadata_parse_array_internal` or around here in `mono_class_create_bounded_array`:

https://github.com/dotnet/runtime/blob/9bfd039f22e92d8ef46384945955470a8dbd80b4/src/mono/mono/metadata/class-init.c#L1173-L1184

I can't tell without running some tests which is the "right" place to change.  I think it's `mono_class_create_bounded_array`, but I'm not certain.